docs/SPEC-composition.md

```diff
@@ -0,0 +1,173 @@
+Flux Composition
+==
+
+The composition capabilities of the flux LSP server enable clients to create a query builder experience without deep knowledge of the flux internals. Composition will create a "Composition-owned" statement in a given flux file, and will be able to add/remove filter calls to that statement. The composition rules apply as follows:
+
+  - Composition must never generate a query that cannot be executed.
+  - Composition should never generate a query that a user with sufficient flux knowledge would never write, e.g. `filter(fn: (r) => r.foo == "bar") |> filter(fn: (r) => r.foo == "bar")`
+  - Composition should never generate a query that results in immediate diagnostic messages.
+  - In the event of conflicts or logical ambiguity, Composition should default to erroring rather than introducing complex assumption logic to attempt recovery.
+
+Given these rules, this document will not attempt to document the resulting flux code generated from each of these requests, as the resulting flux could change as requirements evolve and the corresponding flux would become outdated.
+
+Composition Initialize
+--
+
+The composition initialize request is sent from the client to the server to identify a file that will be used to compose queries. This call must be made before any other Composition requests, otherwise those requests will error. It will result in a `workspace/applyEdit` request from the server to the client with the associated changes to the file. The request can also be used to reset the composition query state by calling it again, and can be considered idempotent.
+
+*Request*
+  - method: `fluxComposition/initialize`
+  - params: `CompositionInitializeParams`
+
+```
+interface CompositionInitializeParams {
+    /*
+     * The text document.
+     */
+    text_document: TextDocumentIdentifier;
+
+    /*
+     * The bucket to initialize the composition query.
+     */
+    bucket: string;
+
+    /*
+     * The measurement to initialize the composition query.
+     */
+    measurement?: string;
+}
+```
+
+*Response*
+  - result: `null`
+  - error: code and message set in case an exception happens during the initialize request.
+
+Add Measurement Filter
+--
+
+The add measurement filter request is sent from the client to the server to add a measurement filter. If a measurement is not already specified, e.g. it was not specified in the initialization and not called previously, it will result in a `workspace/applyEdit` request from the server to the client with the associated filter addition. If a measurement filter already exists, an error will be reported in the form of a `window/showMessage` request from the server to the client with the associated information.
+
+*Request*
+  - method: `fluxComposition/addMeasurementFilter`
```

Comment:
We are now introducing our own methods which are language specific. That breaks the language agnostic purpose of the spec.

The way it works for our specific use case should be fine. But I haven't wrapped my head around any possible future ramifications. I feel like it's a "yes, do it" answer....as long as we accept that the LSP is starting to not be truly open source interoperable. Does this make sense?

Reply:
The LSP spec itself was written to be language agnostic, but with the ability to extend it for language specific purposes. I don't think that violates that. However, while I wrote this with the intent on talking about these interactions as requests, I _think_ they're just commands, in the same way that we have our current injection work.